### PR TITLE
fix(controller): Implement offloading for workflow updates that are re-applied. Fixes #2856

### LINKED ISSUE #2856: Persistence race condition with merging offloaded nodes
If there is a conflict when updating the offloaded nodes for a workflow, there is no merge between the two workflow updates - instead the latter one wins.

This is because the persistUpdates() function is missing support for offloaded nodes when apierr.IsConflict(..) returns true - it needs to make sure that any changes to the nodes in the offload storage are merged with the changes we're trying to save.

Here is an example:

- goroutine A loads workflow + nodes
- goroutine B (which may be in a different process) loads the same workflow + nodes

- they both make changes which mutate the nodes

- goroutine A persists its modified nodes to offload (operator.go:468), let's assume this results in version 'fnv:12', then persists the rest of the workflow metadata to etcd with offload pointing to 'fnv:12'

- goroutine B persists its modified nodes to offload (operator.go:468), let's assume this results in version 'fnv:943', then attempts to persist the workflow to etcd with offload pointing to 'fnv:943'. This fails and it does a merge of the workflow metadata (but not the nodes) in reapplyUpdate, it then tries again and this succeeds.

At no point are the updates to the two sets of nodes merged and therefore the updates to the nodes made by goroutine A which are contained in 'fnv:12' are lost and replaced by the updates in goroutine B.

---

**Message from the maintainers**:

If you are impacted by this bug please add a 👍 reaction to this issue! We often sort issues this way to know what to prioritize.

### PR DESCRIPTION
**Pull request by alexec**
Checklist:

* [x] Either (a) I've created an [enhancement proposal](https://github.com/argoproj/argo/issues/new/choose) and discussed it with the community, (b) this is a bug fix, or (c) this is a chore.
* [x] The title of the PR is (a) [conventional](https://www.conventionalcommits.org/en/v1.0.0/), (b) states what changed, and (c) suffixes the related issues number. E.g. `"fix(controller): Updates such and such. Fixes #1234"`.  
* [x] I've signed the CLA.
* [x] I have written unit and/or e2e tests for my change. PRs without these are unlikely to be merged.
* [x] My builds are green. Try syncing with master if they are not. 
* [x] My organization is added to USERS.md.

### CODE REVIEW

**Comment by alexec** [comment]
* TestPersistErrorWithoutLargeWfSupport
* TestSuspendTemplate
* TestArgoServerSuite/TestWorkflowService/List

File: workflow/controller/operator_persist_test.go [hunk 1]
```diff
@@ -103,7 +100,7 @@ func TestPersistWithLargeWfSupport(t *testing.T) {
 	assert.Empty(t, wf.Status.Nodes)
 	assert.Empty(t, wf.Status.CompressedNodes)
 	// check the updated in-memory version is pre-offloaded state
-	assert.True(t, woc.wf.Status.IsOffloadNodeStatus())
+	assert.False(t, woc.wf.Status.IsOffloadNodeStatus())
```
Comment (alexec) [inline workflow/controller/operator_persist_test.go:103]: behaviour change

File: workflow/controller/operator_persist_test.go [hunk 2]
```diff
@@ -72,15 +72,12 @@ func TestPersistErrorWithoutLargeWfSupport(t *testing.T) {
...
+	controller.offloadNodeStatusRepo, controller.hydrator = getMockDBCtx(errors.New("23324", "test"), false)
 	woc := newWorkflowOperationCtx(wf, controller)
 	woc.operate()
 	wf, err = wfcset.Get(wf.Name, metav1.GetOptions{})
 	assert.NoError(t, err)
-	assert.False(t, wf.Status.IsOffloadNodeStatus())
-	assert.Equal(t, wfv1.NodeRunning, woc.wf.Status.Phase)
-	assert.NotEmpty(t, woc.wf.Status.Nodes)
-	assert.Empty(t, woc.wf.Status.CompressedNodes)
+	assert.Equal(t, wfv1.NodeError, wf.Status.Phase)
```
Comment (alexec) [inline workflow/controller/operator_persist_test.go:80]: corrected behaviour

File: workflow/packer/packer.go [hunk 3]
```diff
@@ -83,7 +84,12 @@ func compressWorkflow(wf *wfv1.Workflow) error {
 		return err
 	}
 	if large {
-		compressedSize, _ := getSize(wf)
+		compressedSize, err := getSize(wf)
+		wf.Status.CompressedNodes = ""
```
Comment (alexec) [inline workflow/packer/packer.go:88]: this was a bug

File: workflow/util/util_test.go [hunk 4]
```diff
@@ -279,91 +276,26 @@ status:
...
-	err = ResumeWorkflow(wfIf, offloadNodeStatusRepo, "suspend", "")
-	assert.NoError(t, err)
-
-	wf, err := wfIf.Get("suspend", metav1.GetOptions{})
-	assert.NoError(t, err)
-	assert.Equal(t, "1234", wf.Status.OffloadNodeStatusVersion)
-}
-
-//workflow is too big but offload is disabled
-func TestResumeWorkflowOffloadDisabled(t *testing.T) {
```
Comment (alexec) [inline workflow/util/util_test.go:332]: these tests are all irrelevant now - they were testing the hydration code by proxy - we don't need that - or want that

**Comment by alexec** [comment]
`TestWorkflowRetryNoPersistence`

File: Makefile [hunk 5]
```diff
@@ -311,8 +311,7 @@ dist/postgres.yaml: $(MANIFESTS) $(E2E_MANIFESTS) $(VERSION_FILE)
 	kustomize build --load_restrictor=none test/e2e/manifests/postgres | sed 's/:$(MANIFESTS_VERSION)/:$(VERSION)/' | sed 's/pns/$(E2E_EXECUTOR)/' > dist/postgres.yaml
 
 dist/no-db.yaml: $(MANIFESTS) $(E2E_MANIFESTS) $(VERSION_FILE)
-	# We additionally disable ALWAYS_OFFLOAD_NODE_STATUS
-	kustomize build --load_restrictor=none test/e2e/manifests/no-db | sed 's/:$(MANIFESTS_VERSION)/:$(VERSION)/' | sed 's/pns/$(E2E_EXECUTOR)/' | sed 's/"true"/"false"/' > dist/no-db.yaml
+	kustomize build --load_restrictor=none test/e2e/manifests/no-db | sed 's/:$(MANIFESTS_VERSION)/:$(VERSION)/' | sed 's/pns/$(E2E_EXECUTOR)/' > dist/no-db.yaml
```
Comment (alexec) [inline Makefile:314]: dead code

File: workflow/controller/operator.go [hunk 6]
```diff
@@ -492,19 +477,22 @@ func (woc *wfOperationCtx) persistUpdates() {
...
 		}
 		woc.wf = wf
 	} else {
 		woc.wf = wf
+		woc.controller.hydrator.HydrateWithNodes(woc.wf, nodes)
+	}
+
+	if !woc.controller.hydrator.IsHydrated(woc.wf) {
+		woc.log.Error("workflow should be hydrated")
+		return
```
Comment (alexec) [inline workflow/controller/operator.go:493]: we should fail here

File: workflow/hydrator/hydrator.go [hunk 7]
```diff
@@ -0,0 +1,79 @@
+package hydrator
```
Comment (alexec) [inline workflow/hydrator/hydrator.go:1]: any thoughts on another name for hydration?

File: workflow/util/util.go [hunk 8]
```diff
@@ -459,7 +409,7 @@ func updateWorkflowNodeByKey(wfIf v1alpha1.WorkflowInterface, repo sqldb.Offload
 			return false, err
 		}
 
-		err = decompressAndFetchOffloadedNodes(wf, repo)
+		err = hydrator.Dehydrate(wf)
```
Comment (alexec) [inline workflow/util/util.go:412]: error: should be hydrate

**Comment by alexec** [comment]
`TestWorkflowRetry`

**Comment by alexec** [comment]
`TestCLIWithServerSuite/TestArchive` timeout

File: workflow/controller/controller.go [hunk 9]
```diff
@@ -482,28 +483,15 @@ func (wfc *WorkflowController) processNextItem() bool {
...
-		woc.wf.Status.Nodes = nodes
-	}
-
-	// Decompress the node if it is compressed
-	err = packer.DecompressWorkflow(woc.wf)
+	err = wfc.hydrator.Hydrate(woc.wf)
 	if err != nil {
-		woc.log.Errorf("workflow decompression failed: %v", err)
+		woc.log.Errorf("hydration failed: %v", err)
 		woc.markWorkflowError(err, true)
```
Comment (mark9white) [inline workflow/controller/controller.go:489]: markWorkflowError tries to access the workflow nodes, which may not exist if the hydration has failed.

File: workflow/controller/operator.go [hunk 10]
```diff
@@ -463,22 +461,9 @@ func (woc *wfOperationCtx) persistUpdates() {
...
-				woc.wf.Status.CompressedNodes = ""
-				woc.wf.Status.OffloadNodeStatusVersion = offloadVersion
-			}
-		}
-	} else if err != nil {
-		woc.log.Warnf("Error compressing workflow: %v", err)
+	err := woc.controller.hydrator.Dehydrate(woc.wf)
+	if err != nil {
+		woc.log.Warnf("Failed to dehydrate: %v", err)
 		woc.markWorkflowError(err, true)
```
Comment (mark9white) [inline workflow/controller/operator.go:467]: Same problem here

File: workflow/controller/operator.go [hunk 11]
```diff
@@ -554,43 +541,56 @@ func (woc *wfOperationCtx) persistWorkflowSizeLimitErr(wfClient v1alpha1.Workflo
 // reapplyUpdate GETs the latest version of the workflow, re-applies the updates and
 // retries the UPDATE multiple times. For reasoning behind this technique, see:
 // https://github.com/kubernetes/community/blob/master/contributors/devel/api-conventions.md#concurrency-control-and-consistency
-func (woc *wfOperationCtx) reapplyUpdate(wfClient v1alpha1.WorkflowInterface) (*wfv1.Workflow, error) {
+func (woc *wfOperationCtx) reapplyUpdate(wfClient v1alpha1.WorkflowInterface, nodes wfv1.Nodes) (*wfv1.Workflow, error) {
```
Comment (mark9white) [inline workflow/controller/operator.go:544]: Could this do with extra unit tests, eg around the multiple attempt handling and merge scenarios?

Comment (alexec) [reply workflow/controller/operator.go:544]: I think you're right about this - toughie to test, but all mockable now.

File: workflow/controller/controller.go (hunk 9 above)
Comment (alexec) [reply workflow/controller/controller.go:489]: catch-22: if we cannot hydrate the workflow, then we cannot access the nodes

File: workflow/controller/operator.go (hunk 10 above)
Comment (alexec) [reply workflow/controller/operator.go:467]: I don't think we need to fix this - I've made changes to `Dehydrate` to ensure that error branches do not result in a dehydrated workflow - could still fail to persist mind you

File: workflow/packer/packer.go [hunk 12]
```diff
@@ -80,10 +81,17 @@ func compressWorkflow(wf *wfv1.Workflow) error {
 	// still too large?
 	large, err := IsLargeWorkflow(wf)
 	if err != nil {
+		wf.Status.CompressedNodes = ""
```
Comment (alexec) [inline workflow/packer/packer.go:84]: an existing bug

**Comment by alexec** [comment]
@mark9white if you're happy that you ran this and it worked, I'll look to get a core team review and appoval.

**Comment by mark9white** [comment]
> @mark9white if you're happy that you ran this and it worked, I'll look to get a core team review and appoval.

That would be great :)

**Comment by alexec** [comment]
@sarabala1979 could I please ask you to review this PR for me?

**Pull request merged**